pkg/descheduler/evictions/evictions.go

```diff
@@ -77,20 +248,137 @@ func NewPodEvictor(
+			cache.ResourceEventHandlerFuncs{
+				AddFunc: func(obj interface{}) {
+					pod, ok := obj.(*v1.Pod)
+					if !ok {
+						klog.ErrorS(nil, "Cannot convert to *v1.Pod", "obj", obj)
+						return
+					}
+					if _, exists := pod.Annotations[EvictionRequestAnnotationKey]; exists {
+						if _, exists := pod.Annotations[EvictionInProgressAnnotationKey]; exists {
+							// Ignore completed/suceeeded or failed pods
+							if pod.Status.Phase != v1.PodSucceeded && pod.Status.Phase != v1.PodFailed {
+								klog.V(3).InfoS("Eviction in background detected. Adding pod to the cache.", "pod", klog.KObj(pod))
+								erCache.addPod(pod)
+							}
+						}
+					}
+				},
+				UpdateFunc: func(oldObj, newObj interface{}) {
+					oldPod, ok := oldObj.(*v1.Pod)
+					if !ok {
+						klog.ErrorS(nil, "Cannot convert oldObj to *v1.Pod", "oldObj", oldObj)
+						return
+					}
+					newPod, ok := newObj.(*v1.Pod)
+					if !ok {
+						klog.ErrorS(nil, "Cannot convert newObj to *v1.Pod", "newObj", newObj)
+						return
+					}
+					// Ignore pod's that are not subject to an eviction in background
+					if _, exists := newPod.Annotations[EvictionRequestAnnotationKey]; !exists {
+						if erCache.hasPod(newPod) {
+							klog.V(3).InfoS("Pod with eviction in background lost annotation. Removing pod from the cache.", "pod", klog.KObj(newPod))
+						}
+						erCache.deletePod(newPod)
+						return
+					}
+					// Remove completed/suceeeded or failed pods from the cache
+					if newPod.Status.Phase == v1.PodSucceeded || newPod.Status.Phase == v1.PodFailed {
+						klog.V(3).InfoS("Pod with eviction in background completed. Removing pod from the cache.", "pod", klog.KObj(newPod))
+						erCache.deletePod(newPod)
+						return
+					}
+					// Ignore any pod that does not have eviction in progress
+					if _, exists := newPod.Annotations[EvictionInProgressAnnotationKey]; !exists {
+						// In case EvictionInProgressAnnotationKey annotation is not present/removed
+						// it's unclear whether the eviction was restarted or terminated.
+						// If the eviction gets restarted the pod needs to be removed from the cache
+						// to allow re-triggering the eviction.
+						if _, exists := oldPod.Annotations[EvictionInProgressAnnotationKey]; !exists {
+							return
+						}
+						// the annotation was removed -> remove the pod from the cache to allow to
+						// request for eviction again. In case the eviction got restarted requesting
+						// the eviction again is expected to be a no-op. In case the eviction
+						// got terminated with no-retry, requesting a new eviction is a normal
+						// operation.
+						klog.V(3).InfoS("Eviction in background canceled (annotation removed). Removing pod from the cache.", "annotation", EvictionInProgressAnnotationKey, "pod", klog.KObj(newPod))
+						erCache.deletePod(newPod)
+						return
+					}
+					// Pick up the eviction in progress
+					erCache.addPod(newPod)
```

Comment:
We should prob log this operations for consistency reasons (we do for the rest of add/delete operations).

Also same concern about the assume operation.

Reply:
and should be guarded by `hasPod`

Reply:
Updated